internal/servicelog/ringbuffer.go

```diff
@@ -0,0 +1,369 @@
+// Copyright (c) 2021 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package servicelog
+
+import (
+	"errors"
+	"io"
+	"sync"
+)
+
+var (
+	ErrRange = errors.New("out of range")
+)
+
+type RingPos int64
+
+// RingBuffer is a io.Writer that uses a single byte buffer to store data written to it
+// until Release is called on the range no-longer required. RingBuffer is effectively a
+// linear allocator with sequential frees that must be done in the same order as the
+// allocations.
+type RingBuffer struct {
+	rwlock      sync.RWMutex
+	readIndex   RingPos
+	writeIndex  RingPos
+	writeClosed bool
+	data        []byte
+
+	iteratorMutex sync.RWMutex
+	iteratorList  []*iterator
+}
+
+var _ io.WriteCloser = (*RingBuffer)(nil)
+
+// NewRingBuffer creates a RingBuffer with the provided size in bytes for the backing
+// buffer.
+func NewRingBuffer(size int) *RingBuffer {
+	rb := RingBuffer{
+		data: make([]byte, size),
+	}
+	return &rb
+}
+
+// Close closes the writer to further writes, readers may continue.
+func (rb *RingBuffer) Close() error {
+	rb.rwlock.Lock()
+	defer rb.rwlock.Unlock()
+	if rb.writeClosed {
+		return nil
+	}
+	rb.writeClosed = true
+	rb.releaseIterators()
+	return nil
+}
+
+// Closed returns true if the writing side has closed.
+func (rb *RingBuffer) Closed() bool {
+	rb.rwlock.RLock()
+	defer rb.rwlock.RUnlock()
+	return rb.writeClosed
+}
+
+// Write writes p to the backing buffer, allocating the number of bytes in p.
+// If p is larger than the size of the buffer then io.ErrShortWrite is returned and the
+// number of bytes written. If the p is larger than the number of bytes available,
+// then the tail is discarded to make room.
+func (rb *RingBuffer) Write(p []byte) (written int, _ error) {
```

Comment:
Can we please have `err` as the second variable name?  This is reading incorrect to me right now as I generally associate the `_` with "unused", but that error may actually be set to a valid value below and should be checked by the call site.